### PR TITLE
Instances of dataclasses are preserved when convert is ConvertMode.ALL

### PR DESCRIPTION
**Pull request by guillaumegenthial**
## Motivation

Draft to resolve https://github.com/facebookresearch/hydra/issues/1183.

DISCLAIMER : this is a DRAFT and should not be merged. This PR is merely a POC of a potential solution.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

**Comment by omry** [comment]
Hi @guillaumegenthial, thanks for giving it a shot.
I actually just did this last nigh (took a minor bugfix to OmegaConf as well) and I think my implementation is cleaner.
Give it a shot, I am not merging it yet because I killed the test for live interpolation there and I want to reconsider, but otherwise I think it's good.

File: tests/test_utils.py [hunk 1]
```diff
@@ -987,23 +987,23 @@ def test_convert_params(input_: Any, expected: Any, convert_mode: Any):
...
-    # Even in ConvertMode.ALL - the underlying dict and list are converted to DictConfig and ListConfig
-    # because the parent DictConfig cannot hold unwrapped list and dict.
-    # This behavior is unique to instantiated objects that are dataclasses.
+    """Test instantiation of dataclass instances.DictConfig
 
+    Except for when convert = ConvertMode.ALL, dataclass instances are
+    treated as OmegaConf DictConfig.
+    """
     cfg = OmegaConf.create(input_)
-    kwargs = {"a": {"_convert_": convert}}
```
Comment (guillaumegenthial) [inline tests/test_utils.py:998]: If `_convert_` is only applied to `"a"`, it does not work (currently).

The question is : if a top-level node has `convert = convert.None`, does it supersede all children convert modes?

File: hydra/utils.py [hunk 2]
```diff
@@ -33,6 +34,36 @@ class ConvertMode(Enum):
     ALL = 2
 
 
+def unbox(obj: Any, convert: ConvertMode):
+    """Unbox object."""
```
Comment (guillaumegenthial) [inline hydra/utils.py:38]: Convoluted, I try to remove all OmegaConf containers when convert mode is ALL.

File: hydra/utils.py [hunk 3]
```diff
@@ -33,6 +34,36 @@ class ConvertMode(Enum):
...
+            d[key] = unbox(value, convert)
+        result = d
+    elif isinstance(obj, list):
+        result = [unbox(item, convert) for item in obj]
+    elif OmegaConf.is_list(obj):
+        lst = [] if convert == ConvertMode.ALL else OmegaConf.create([], flags={"allow_objects": True})
+        for it in obj:
+            lst.append(unbox(it, convert))
+        result = lst
+    elif convert != ConvertMode.ALL:
```
Comment (guillaumegenthial) [inline hydra/utils.py:55]: I add this for the PARTIAL mode -> I take the data class instance and convert it back to a structured config

**Comment by guillaumegenthial** [comment]
Closing, duplicate of #1223

**Pull request closed without merging**